### PR TITLE
Introduce autoFQTag

### PR DESCRIPTION
**Pull request by masatake**
See #1260 about the use case.

### CODE REVIEW

**Comment by masatake** [comment]
@bitterjug, about autoFQTag, my side is ready:-)

File: docs/optlib.rst [hunk 1]
```diff
@@ -378,6 +379,52 @@ to `TRUE`. See `cork API`.
...
+Generating fully qualified tags automatically from scope information
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. NOT REVIEWED YET
+
+If scope fields are filled properly with `{scope=...}` regex flags,
+you can use the field values for generating fully qualified tags.
+
+Specify `{autoFQTag}` to the end of ``--langdef=<LANG>`` option like
+``-langdef=Foo{autoFQTag}`` to make ctags generate fully qualified
```
[the remark this answers is not in the document]
Comment (masatake) [reply docs/optlib.rst]: Thank you.

File: docs/optlib.rst [hunk 2]
```diff
@@ -378,6 +379,52 @@ to `TRUE`. See `cork API`.
 NOTE: This flag doesn't work well with ``--mline-regex-<LANG>=``.
 
 
+Generating fully qualified tags automatically from scope information
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. NOT REVIEWED YET
+
+If scope fields are filled properly with `{scope=...}` regex flags,
+you can use the field values for generating fully qualified tags.
```
[the remark this answers is not in the document]
Comment (masatake) [reply docs/optlib.rst]: The scope stack is for filling scope field of tags.
`{scope=...}` flag is for manipulating the scope stack.
`{scope=ref}` is special. It doesn't change the state of the stack. However,
it fill the scope field of a tag wit the value at the top of stack.


Quoted from ctags-optlib.7.rst.in:

> You  can  fill  the  scope  field  of captured tag with
>{scope=ref}.  If  {scope=ref}  flag  is  given,   ctags
>attaches  the  tag  at the top to the tag captured with
> --regex-<LANG> as the value for the scope: field.

Do I make sense?

**Comment by bitterjug** [comment]
You make sense. Apologies for not Reading The ... Manual!

Let me check:
Is it correct that the fully qualified name only appears if you have used scope=ref? And internally does it use the scope _field_ set by scope=ref when generating the <scope>.<id> name? 

My intuition for how this works was incorrect. I had assumed you would use the entier contents of the scope stack, with dots between them, as the scope. But now I infer we use only whatever identifier was taken from the top of the stack with scope=ref. And if scope=ref was not used, then no FQ tag will be generated.

**Comment by masatake** [comment]
> Let me check:
Is it correct that the fully qualified name only appears if you have used scope=ref? And internally does it use the scope field set by scope=ref when generating the . name?

Yes. I will explain the mechanism in my words.
If you use scope= in your optlib parser, ctags gives a sequential id each tag. We call it corkIndex.
See http://docs.ctags.io/en/latest/internal.html#cork-api for more details.
What pushed to the scope stack is corkIndex of a tag.
When using scope=ref, a corkIndex at the top of stack is put to the scope field of a tag.
If a tag is given, you can get upper scope recursively by referring the field of the tag.

>  My  intuition for how this works was incorrect. I had assumed you would use  the entier contents of the scope stack, with dots between them, as the  scope. But now I infer we use only whatever identifier was taken from  the top of the stack with scope=ref. And if scope=ref was not used, then  no FQ tag will be generated.

A bit incorrect. What pushed to the stack is an id for a tag. The id (integer) can be use do access a data structure represeting a tag.
... I will write a figure.

**Comment by bitterjug** [comment]
Thanks for explaining that. I wonder if a user of Optlib should be required
to think about cork indexes and scope fields.
I think, from the perspective of documentation, it would suffice to speak
only of the stack. I suggest this because the
actual optlib options refer to operations like "push" and "pop" and even
"refer" which, if you're a forth programmer,
might be a reference to the "thing" on the stack. If we can write good
instructions without introducing these concepts
I think it will be easier to learn and use. Though a figure might help to
explain the truth about the internals, it think
the documentations would do better to explain to an optlib user how to
achieve what they want to with the fewest possible
number of internal concepts being exposed to them.

What do you think?

**Comment by masatake** [comment]
scope field is the concept of tags file. It is not newly introduced in Universal-ctags.
Look at the end of "TAG FILE FORMAT" of "man ctags" of exuberant-crags.
The man page uses the term "field" explicitly. So we should update the man page and try the users understand the concept of "field" and "scope field". I would like to assume "scope filed" as the base knowledge that optlib developers must have.

The newly concept introduced in Universal-ctags is only scope stack.
So what the optlib developer must have to know for filling scope field and emiting FQ tags are:
* the concept of scope field itself, and
* scope stack operation.
If I wrote the documents properly, autoFQtag regex flag can be understood based on the above two things.

The scope field is what you can control with `--fields=` option.
Let show me an example:

foo.c:
```
struct point {
 int x, y;
};
```
output:
```
[yamato@master]~/var/ctags-github% ./ctags -o - /tmp/foo.c            
point	/tmp/foo.c	/^struct point {$/;"	s	file:
x	/tmp/foo.c	/^	int x, y;$/;"	m	struct:point	typeref:typename:int	file:
y	/tmp/foo.c	/^	int x, y;$/;"	m	struct:point	typeref:typename:int	file:
[yamato@master]~/var/ctags-github% ./ctags --fields=-s -o - /tmp/foo.c  
point	/tmp/foo.c	/^struct point {$/;"	s	file:
x	/tmp/foo.c	/^	int x, y;$/;"	m	typeref:typename:int	file:
y	/tmp/foo.c	/^	int x, y;$/;"	m	typeref:typename:int	file:
[yamato@master]~/var/ctags-github% ./ctags --fields=+Z -o - /tmp/foo.c 
point	/tmp/foo.c	/^struct point {$/;"	s	file:
x	/tmp/foo.c	/^	int x, y;$/;"	m	scope:struct:point	typeref:typename:int	file:
y	/tmp/foo.c	/^	int x, y;$/;"	m	scope:struct:point	typeref:typename:int	file:
```

By default, scope values are printed but the scope names (prefix, `scope:`) are omitted.
I think this is bad choice in Exuberant-ctag development. So I added `Z` as a field flag controlling whether printing the field name, `scope:` or not.

**Comment by ahakanbaba** [comment]
@masatake this PR has merge conflicts. Do you think you can resolve them first ? 

After that, I am more than happy to write test cases related to #1916  or additional documentation. 

I need some direction for writing documentation too. Where do you want the docs and also what should be documented.

**Comment by masatake** [comment]
@ahakanbaba, what I would like to see is 16fc5cf.

**Comment by ahakanbaba** [comment]
Sorry @masatake  I do not understand your comment. I see that you have written some documentation. In what way can I help ?

**Comment by masatake** [comment]
@ahakanbaba, I would like you to review the section added in 16fc5cf. Does the section make sense? Don't I use too broken Engrish?

**Comment by ahakanbaba** [comment]
Thanks @masatake  it took me a while to catch up with the discussion in  #1260 and in here. 

I made some comments at your doc at the [`16fc5cf`](https://github.com/universal-ctags/ctags/commit/16fc5cfa001e6b2aafc657bc9a43ee6e30a56225) directly.  (Maybe I should have added my commits to this PR directly) Next time I will do that.

**Comment by ahakanbaba** [comment]
Thanks for your responses @masatake 
I have responded to your comments on [16fc5cf](https://github.com/universal-ctags/ctags/commit/16fc5cfa001e6b2aafc657bc9a43ee6e30a56225)

This PR looks mergeable to me.

**Comment by masatake** [comment]
I will work on #1903 before working on #1916.

**Comment by masatake** [comment]
@ahakanbaba, thank you.

**Pull request merged**